crates/kas-core/src/event/cx/key.rs

```diff
@@ -201,42 +192,126 @@ impl EventState {
         }
     }
 
-    /// End Input Method Editor focus on `target`, if present
-    #[inline]
-    pub fn cancel_ime_focus(&mut self, target: Id) {
-        if let Some(pending) = self.pending_sel_focus.as_mut() {
-            if pending.target.as_ref() == Some(&target) {
-                pending.ime = None;
-            }
-        } else if self.ime.is_some() && self.sel_focus.as_ref() == Some(&target) {
-            self.pending_sel_focus = Some(PendingSelFocus {
-                target: Some(target),
-                key_focus: self.key_focus,
-                ime: None,
-                source: FocusSource::Synthetic,
-            });
-        }
-    }
-
     /// Set IME cursor area
     ///
-    /// This should be called after receiving [`Event::ImeFocus`], and any time
-    /// that the `rect` parameter changes, until [`Event::LostImeFocus`] is
-    /// received.
+    /// This should be called after receiving
+    /// <code>[Event::Ime][]([Ime::Enabled][crate::event::Ime::Enabled])</code>,
+    /// and any time that the `rect` parameter changes, until
+    /// <code>[Event::Ime][]([Ime::Disabled][crate::event::Ime::Disabled])</code>
+    /// is received.
     ///
     /// This sets the text cursor's area, `rect`, relative to the widget's own
     /// coordinate space. If never called, then the widget's whole rect is used.
     ///
     /// This does nothing if `target` does not have IME-enabled input focus.
     #[inline]
     pub fn set_ime_cursor_area(&mut self, target: &Id, rect: Rect) {
-        if self.ime.is_some() && self.sel_focus.as_ref() == Some(target) {
+        if self.ime_is_enabled && self.sel_focus.as_ref() == Some(target) {
             self.ime_cursor_area = rect;
         }
     }
 }
 
 impl<'a> EventCx<'a> {
+    /// Request IME focus
+    ///
+    /// IME focus requires selection focus (see
+    /// [`EventState::request_key_focus`], [`EventState::request_sel_focus`]).
+    /// The request fails if this has not already been obtained (wait for
+    /// [`Event::SelFocus`] or [`Event::KeyFocus`] before calling this method).
+    ///
+    /// To support Input Method Editors, call this method after receiving
+    /// selection focus and handle [`Event::Ime`] events.
+    ///
+    /// If `target` does not have selection focus this request will be ignored.
+    /// If `target` already has IME focus, it will be replaced (using the new
+    /// parameters provided).
+    pub fn request_ime_focus(
+        &mut self,
+        target: Id,
+        hint: ImeHint,
+        purpose: ImePurpose,
+        mut surrounding_text: Option<ImeSurroundingText>,
+    ) {
```

Comment:
All `IME` requests except `set_ime_cursor_area` are now on `EventCx` and require pre-existing selection focus. (All requests require the `target` be specified and ignore the request if this target does not have selection focus.)